### ISSUE TITLE
Error: WebSocket is already in CLOSING or CLOSED state

### ISSUE THREAD

**Issue by Mangatt** [enisdenjo/graphql-ws#331]
After switching from `subscriptions-transport-ws` in Apollo client, there is always error `WebSocket is already in CLOSING or CLOSED state` after initialization in browser console.

**Further Information**
- graphql-ws: 5.6.2
- @apollo/client: 3.5.10
- Chrome: 99.0.4844.51

```js
const wsLink = new GraphQLWsLink(
	createClient({
		url: 'ws://localhost:4000',
		connectionParams: async () => ({
			credentials: 'include',
			authorization: await getAuthToken(),
		}),
	})
)
```

**Comment by enisdenjo** [comment]
Does it work without using the `GraphQLWsLink`? Create a client and use it directly through `client.subscribe`.

If there's no problem then, consider opening an issue in [`apollo-client`](https://github.com/apollographql/apollo-client) instead.

However, if it's still breaking, please provide more info - the full callstack of the Error, or ideally a minimal repro.

**Comment by Mangatt** [comment]
Callstack:
```
WebSocket is already in CLOSING or CLOSED state.
socket.onopen @ client.mjs:167
```

**Comment by Mangatt** [comment]
And yes, error is present event without `GraphQLWsLink`, when calling `client.subscribe()`.

**Comment by Mangatt** [comment]
Just a thought - can this be possibly connected to async `connectionParams`? Because as of now I can get some subscriptions without error. Error appears randomly after browser reload, without changing anything in code.

**Comment by enisdenjo** [comment]
> Just a thought - can this be possibly connected to async connectionParams? 

[The connection process stalls until the `connectionParams` resolve](https://github.com/enisdenjo/graphql-ws/blob/126a19abf374021e9899b40b52824dfcabf666bf/src/client.ts#L627-L630), it could be that you're unsubscribing before `connectionParams`. This might cause the socket to close early. Can you please check that?

It would be of great help if you could offer a minimal repro.

**Comment by enisdenjo** [comment]
Oh and, does your `connectionParams` even resolve? Does it maybe throw instead?
```diff
const wsLink = new GraphQLWsLink(
	createClient({
		url: 'ws://localhost:4000',
		connectionParams: async () => ({
			credentials: 'include',
			authorization: await getAuthToken(),
		}),
+               on: {
+                 error: (err) => {
+                   console.log(err); // 👈 does this log?
+                 },
+               },
	})
)
```

**Comment by enisdenjo** [comment]
Inspecting the [callstack](https://github.com/enisdenjo/graphql-ws/blob/126a19abf374021e9899b40b52824dfcabf666bf/src/client.ts#L631) and taking a hint from your [further comment](https://github.com/enisdenjo/graphql-ws/issues/331#issuecomment-1066123926) - I suspect that the `connectionParams` is taking too long to resolve, causing the server to kick off the client respecting the [`connectionInitWaitTimeout`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#connectioninitwaittimeout).

To solve it, simply increase the timeout on the server or speed up the `connectionParams`.

The fix released in [v5.6.3](https://github.com/enisdenjo/graphql-ws/releases/tag/v5.6.3) stops execution after `connectionParams` if the connection was closed while waiting. Previously, the client was trying to send a message afterwards regardless, leading to the error in the browser's console.

**Comment by Mangatt** [comment]
Thanks, 5.6.3 have solved the problem.
